✨ NEW: Add superscript plugin and tests

elijahgreenstein opened the pull request:
Add a superscript plugin and associated tests ported from <https://github.com/markdown-it/markdown-it-sup>.

Update documentation with superscript plugin and remove superscript (and subscript, added in PR <https://github.com/executablebooks/mdit-py-plugins/pull/122>) from list of plugins to port.

elijahgreenstein committed:
Add superscript plugin

Plugin ported from https://github.com/markdown-it/markdown-it-sup

elijahgreenstein committed:
Add tests for superscript plugin

Fixtures file is copied from the `markdown-it-sup` tests:
https://github.com/markdown-it/markdown-it-sup/blob/master/test/fixtures/sup.txt

Test is modelled on other Markdown-It-Py inline tests.

elijahgreenstein committed:
Add superscript plugin to documentation

Also, remove superscript (and subscript) from list of plugins to port.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

elijahgreenstein committed:
Remove unnecessary import

minchinweb commented:
Any chance of this being merged soon? I'd like to use it for my projects...

KyleKing commented on tests/test_superscript.py:
```diff
@@ -0,0 +1,23 @@
+        md.disable("code")
+    md.options["xhtmlOut"] = False
+    text = md.render(input)
+    print(text)
```
This shouldn't be necessary because pytest will output the error

```suggestion
```

KyleKing commented on tests/test_superscript.py:
```diff
@@ -0,0 +1,23 @@
+def test_superscript_fixtures(line, title, input, expected):
+    md = MarkdownIt("commonmark").use(superscript_plugin)
+    if "DISABLE-CODEBLOCKS" in title:
+        md.disable("code")
```
The test script doesn't have this text, so you don't need to include this check:

```suggestion
```

KyleKing commented on mdit_py_plugins/superscript/index.py:
```diff
@@ -0,0 +1,108 @@
+from markdown_it.rules_inline import StateInline
+
+UNESCAPE_RE = re.compile(r"\\([ \\!\"#$%&'()*+,./:;<=>?@[\]^_`{|}~-])")
+WHITESPACE_RE = re.compile(r"(^|[^\\])(\\\\)*\s")
```
While `WHITESPACE_RE` is the same, I'm not sure if `UNESCAPE_RE` should diverge the way it does and maybe they could be standardized and extracted into a utility file?

https://github.com/elijahgreenstein/mdit-py-plugins/blob/e2fb671b3a66e86690239137429d5ef69f5e06d5/mdit_py_plugins/subscript/__init__.py#L25-L26

elijahgreenstein replied:
Hey there, thanks for the help cleaning up the tests! As for `UNESCAPE_RE`, I checked the original JS [subscript plugin](https://github.com/markdown-it/markdown-it-sub/blob/master/index.mjs#L4) and the original JS [superscript plugin](https://github.com/markdown-it/markdown-it-sup/blob/master/index.mjs#L4), and it looks like the Python port of the subscript plugin here (in `mdit_py_plugins`) included an extra backslash in the middle of the `UNESCAPE_RE` string. As you suggested, I've standardized the expressions for both plugins and moved them into the existing utility file.

KyleKing replied:
Great!

KyleKing reviewed (approved):
Looks just like the JS version, thank you!

I'm not a maintainer, so I don't know if my review will help close this, but it can't hurt 👍🏻

elijahgreenstein committed:
Remove unnecessary print

Co-authored-by: Kyle King <KyleKing@users.noreply.github.com>

elijahgreenstein committed:
Remove unnecessary title check

Co-authored-by: Kyle King <KyleKing@users.noreply.github.com>

elijahgreenstein committed:
Move shared sub/superscript regex to utility file

The subscript and superscript plugins both use the same regular
expressions. This commit standardizes and centralizes the regex into a
utility file.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

elijahgreenstein committed:
Correct docstring regarding use of spaces

My original docstring stated that superscripted text cannot contain any
whitespace characters. In fact, superscripted text can contain space
characters, as long as they are escaped.

minchinweb commented:
As an interim measure, I've released a standalone version of the plugin so I could use it locally --> https://github.com/minchinweb/minchin.md-it.superscript

chrisjsewell committed:
Merge branch 'master' into superscript

chrisjsewell reviewed (approved):
all good cheers
